### PR TITLE
chore: add VirtualizedList  best practices

### PR DESCRIPTION
**Pull request by MounirDhahri**
### Description

This PR adds some of the best practices to improve lists performance on Eigen. Most of the items here do also apply for other project, I should may create a blog post about this.

**Context:** https://artsy.slack.com/archives/C02BAQ5K7/p1707489650981399

### PR Checklist

- [x] I have tested my changes on **iOS** and **Android**.
- [x] I hid my changes behind a **[feature flag]**, or they don't need one.
- [x] I have included **screenshots** or **videos**, or I have not changed the UI.
- [x] I have added **tests**, or my changes don't require any.
- [x] I added an **[app state migration]**, or my changes do not require one.
- [x] I have documented any **follow-up work** that this PR will require, or it does not require any.
- [x] I have added a **changelog entry** below, or my changes do not require one.

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to **run** this PR on the simulator or device.

<details><summary>Changelog updates</summary>

#### Cross-platform user-facing changes

-

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

- Virtualized lists best practices - mounir

</details>

Need help with something? Have a look at our [docs], or get in touch with us.

[app state migration]: ../blob/main/docs/adding_state_migrations.md
[feature flag]: ../blob/main/docs/developing_a_feature.md
[docs]: ../blob/main/docs/README.md

### CODE REVIEW

**Comment by ArtsyOpenSource** [comment]
### This PR contains the following changes:

- Dev changes (Virtualized lists best practices - mounir)
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against b7bbb16e6caf395d8850de79859023a76715cabd
</p>

File: docs/best_practices.md [hunk 1]
```diff
@@ -170,6 +178,102 @@ For the most part you don't have to worry about this.
 
 In React-native, we use react-tracking as a wrapper for the tracking events we send to Segment. You can read more about the implementation [here](./analytics_and_tracking.md).
 
+### VirtualizedList best practices
+
+Smoothly rendering lists and animations is crucial for a positive user experience. However, performance issues can arise, causing lag, stuttering, and decreased responsiveness. Below are few tips to achieve that in Eigen:
+
+#### Never nest ScrollViews. If you feel like there is no other way, it's probably a better idea to talk to the designer to adjust the approach they're suggested instead. Our `Screen` wrappers from `palette-mobile` expose performant `ScrollView` based components such as `Flatlist` and `Flashlist` that can be used and would save you the nesting.
+
+#### Always default to `Flashlist`. Think of it as `Flatlist` on steroids.
```
Comment (damassi) [inline docs/best_practices.md:187]: > default to `Flashlist`
Link or small example just below?

File: docs/best_practices.md [hunk 2]
```diff
@@ -170,6 +178,102 @@ For the most part you don't have to worry about this.
...
+
+> **What if you followed all the above steps and you still had performance issues:?**
```
Comment (damassi) [inline docs/best_practices.md:189]:
```suggestion
> **What if you followed all the above steps and still have performance issues?**
```

File: docs/best_practices.md (hunk 1 above)
Comment (MounirDhahri) [reply docs/best_practices.md]: good idea

File: docs/best_practices.md [hunk 3]
```diff
@@ -148,35 +156,135 @@ Refactoring components using HoCs in favour of hooks is encouraged.
...
+
+If you feel like there is no other way, it's probably a better idea to talk to the designer to adjust the approach they're suggested instead. Our `Screen` wrappers from `palette-mobile` expose performant `ScrollView` based components such as `Flatlist` and `Flashlist` that can be used and would save you the nesting.
+
+### Always default to `Flashlist`.
+
+Think of `Flashlist` as `Flatlist` on steroids. It's fast, performant and easy to use.
+
+> **What if you followed all the above steps and you still had performance issues:?**
+
+Below are some tips to improve the performance further, don't follow them UNLESS you need to, premature optimisation will haunt you and can lead to issues that are not trivial to debug (plus it's arguably not the best use of your time)
```
Comment (damassi) [inline docs/best_practices.md:195]:
```suggestion
Below are some tips to improve the performance further; don't follow them UNLESS you need to, as premature optimization will haunt you and can lead to issues that are non-trivial to debug (plus it's arguably not the best use of your time).
```

File: docs/best_practices.md [hunk 4]
```diff
@@ -148,35 +156,135 @@ Refactoring components using HoCs in favour of hooks is encouraged.
...
+
+### Use `memo` for the rescue. See: https://reactnative.dev/docs/optimizing-flatlist-configuration#use-memo
```
Comment (damassi) [inline docs/best_practices.md:197]:
```suggestion
### Use `memo` to the rescue. See: https://reactnative.dev/docs/optimizing-flatlist-configuration#use-memo
```

File: docs/best_practices.md [hunk 5]
```diff
@@ -148,35 +156,135 @@ Refactoring components using HoCs in favour of hooks is encouraged.
...
+```
+
+</details>
+
+### Does your component contain animations?
+
+<details>
+<summary>If yes, consider moving all the animations to the native thread, how?</summary>
+
+Example of a component implementing a fade in animation with Moti.
```
Comment (damassi) [inline docs/best_practices.md:236]:
```suggestion
Example of a potentially problematic component implementing a fade in animation with Moti.
```

**Review by damassi — DISMISSED** [review]
Looks great @MounirDhahri 👍 

One little question I have is: In the solution at the bottom, where we drop down into vanilla reanimated, do we _have_ to use it, or would that apply to a moti `AnimatedView` as well? Because if we can keep the better API that would be ideal, though I also see the value in suggesting a more bare-metal approach when the question itself has to do with performance. (No blockers or anything here, mostly just curious!)

**Comment by MounirDhahri** [comment]
@damassi yes, maybe better to go with `Moti` there as you suggested because it will work as well. I will update

**Review by gkartalis — DISMISSED** [review]
thank you!

**Review by MrSltun — DISMISSED** [review]
Thank you Mounir!

File: docs/best_practices.md [hunk 6]
```diff
@@ -148,35 +156,164 @@ Refactoring components using HoCs in favour of hooks is encouraged.
...
+
+Smoothly rendering lists and animations is crucial for a positive user experience. However, performance issues can arise, causing lag, stuttering, and decreased responsiveness. Below are few tips to achieve that in Eigen:
+
+### Never nest ScrollViews.
+
+If you feel like there is no other way, it's probably a better idea to talk to the designer to adjust the approach they're suggested instead. Our `Screen` wrappers from `palette-mobile` expose performant `ScrollView` based components such as `Flatlist` and `Flashlist` that can be used and would save you the nesting.
+
+### Always default to `FlashList`.
+
+Think of `FlashList` as `Flatlist` on steroids. It's fast, performant and easy to use.
```
Comment (olerichter00) [inline docs/best_practices.md:191]: 🙃

File: docs/best_practices.md [hunk 7]
```diff
@@ -148,35 +156,164 @@ Refactoring components using HoCs in favour of hooks is encouraged.
...
+}
+```
+
+</details>
+```
+```
+
+> **What if you followed all the above steps and still have performance issues?**
+
+Below are some tips to improve the performance further; don't follow them UNLESS you need to, as premature optimization will haunt you and can lead to issues that are non-trivial to debug (plus it's arguably not the best use of your time).
```
Comment (olerichter00) [inline docs/best_practices.md:218]: 👏

**Review by olerichter00 — DISMISSED** [review]
Thank you 🌟

**Pull request merged**